node/src/components/contract_runtime/rewards.rs

```diff
@@ -0,0 +1,380 @@
+    ) -> Result<Ratio<U512>, RewardsError> {
+        if era_id.is_genesis() {
+            return Ok(Ratio::zero());
+        }
+
+        let era = self
+            .0
+            .get(&era_id)
+            .ok_or(RewardsError::EraIdNotInEraRange(era_id))?;
+        let weight = era
+            .weights
+            .get(validator)
+            .ok_or_else(|| RewardsError::ValidatorKeyNotInEra(Box::new(validator.clone())))?;
+
+        Ok(Ratio::new(*weight, era.total_weights))
+    }
+
+    /// For debugging purpose.
+    fn list_of_eras_id(&self) -> Vec<u64> {
+        self.0.keys().map(|id| id.value()).collect()
+    }
+}
+
+pub(crate) async fn rewards_for_era<REv: ReactorEventT>(
+    effect_builder: EffectBuilder<REv>,
+    era_id: EraId,
+    switch_block_height: u64,
+    chainspec: Arc<Chainspec>,
+) -> Result<BTreeMap<PublicKey, U512>, RewardsError> {
+    fn increase_value_for_key(
+        map: &mut BTreeMap<PublicKey, Ratio<U512>>,
+        key: PublicKey,
+        value: Ratio<U512>,
+    ) {
+        // We only want to include non-zero values into the rewards, because it does not make a lot
+        // of sense to mint and transfer a reward of 0.
+        // This typically concerns the producer of genesis, for example, which is the system.
+        if value.numer() != &U512::zero() {
+            map.entry(key)
+                .and_modify(|amount| *amount += value)
+                .or_insert(value);
+        }
+    }
+
+    fn to_ratio_u512(ratio: Ratio<u64>) -> Ratio<U512> {
+        Ratio::new(U512::from(*ratio.numer()), U512::from(*ratio.denom()))
+    }
+
+    // All the blocks that may appear as a signed block. They are collected upfront, so that we
+    // don't have to worry about doing it one by one later.
+    //
+    // They are sorted from the oldest to the newest:
+    let cited_blocks = {
+        let cited_block_height_start = if era_id.is_genesis() {
+            0
+        } else {
+            let previous_era_id = era_id.saturating_sub(1);
+            let previous_era_switch_block_header = effect_builder
+                .get_switch_block_header_by_era_id_from_storage(previous_era_id)
+                .await
+                .ok_or(RewardsError::MissingSwitchBlock(previous_era_id))?;
+
+            previous_era_switch_block_header
+                .height()
+                .saturating_sub(chainspec.core_config.signature_rewards_max_delay)
+        };
+        let range_to_fetch = cited_block_height_start..switch_block_height;
+
+        tracing::info!(
+            "[era {}] block range to fetch: {:?}",
+            era_id.value(),
+            range_to_fetch
+        );
+
+        collect_past_blocks_batched(effect_builder, range_to_fetch).await
+    };
+
+    //tracing::info!("blocks actually fetched: {cited_blocks:?}");
+    tracing::info!(
+        "[era {}] blocks fetched: {}/{}",
+        era_id.value(),
+        cited_blocks.iter().flatten().count(),
+        cited_blocks.len(),
+    );
```

Comment:
Let's try to keep to using structured logging, for example like this:
```suggestion
    tracing::info!(
        era_id = %era_id.value(),
        num_fetched_blocks = %cited_blocks.iter().flatten().count(),
        num_requested_blocks = %cited_blocks.len(),
        "rewards: blocks fetched",
    );
```

Reply:
OK, I didn't know how it works. I even think we can get rid of `rewards:` because it's already in the file name.